### PR TITLE
EVM: respect vm.prank during calls to create

### PR DESCRIPTION
**Pull request by d-xo**
## Description

Updates our `prank` implementation to match foundrys. `prank` is now respected during contract creation. 

## Checklist

- [x] tested locally
- [x] added automated tests
- [ ] updated the docs
- [x] updated the changelog

### CODE REVIEW

File: test/test.hs [hunk 1]
```diff
@@ -1151,6 +1151,35 @@ tests = testGroup "hevm"
...
+                function f() external {
+                  Vm vm = Vm(0x7109709ECfa91a80626fF3989D68f67F5b1DD12D);
+
+                  Owned target = new Owned();
+                  assert(target.owner() == address(this));
+
+                  address usr = address(1312);
+                  vm.prank(usr);
+                  target = new Owned();
+                  assert(target.owner() == usr);
```
Comment (d-xo) [inline test/test.hs:1176]: We should check that it changes back afterwards...

File: src/EVM.hs [hunk 2]
```diff
@@ -834,13 +834,18 @@ exec1 = do
...
                 availableGas <- use (#state % #gas)
-                let
-                  (cost, gas') = costOfCreate fees availableGas xSize False
-                newAddr <- createAddress self this.nonce
+                let (cost, gas') = costOfCreate fees availableGas xSize False
+
+                -- handle `prank`
+                let from' = fromMaybe self vm.config.overrideCaller
+                resetCaller <- use (#config % #resetCaller)
+                when (resetCaller) $ assign (#config % #overrideCaller) Nothing
```
Comment (blishko) [inline src/EVM.hs:842]: Can you explain for me what these two lines do?

**Review by msooseth — APPROVED** [review]
lgtm

**Comment by blishko** [comment]
I think the current state of this PR is that the original issue has been fixed, but the `badVault` test still exhibits some unexpected behaviour. Is that right, @d-xo?

File: src/EVM.hs (hunk 2 above)
Comment (msooseth) [reply src/EVM.hs]: These are no longer there. But they reset the caller after `prank`. Only when `startPrank` should the caller remain the pranked caller. I am closing because this has been fixed as part of the #240 issue. Let me know if you have any questions, I'd be happy to answer. It's perhaps worthwhile reading the associated PR that closed that issue.

**Comment by msooseth** [comment]
This should now pass all the way through. Can you please review @elopez and/or @blishko ?;

File: src/EVM.hs [hunk 3]
```diff
@@ -852,13 +852,18 @@ exec1 = do
...
                 availableGas <- use (#state % #gas)
-                let
-                  (cost, gas') = costOfCreate fees availableGas xSize False
-                newAddr <- createAddress self this.nonce
+                let (cost, gas') = costOfCreate fees availableGas xSize False
+
+                -- handle `prank`
+                let from' = fromMaybe self vm.state.overrideCaller
+                resetCaller <- use (#state % #resetCaller)
+                when resetCaller $ assign (#state % #overrideCaller) Nothing
```
Comment (elopez) [inline src/EVM.hs:860]: just for consistency

```suggestion
                when resetCaller $ do
                  assign (#state % #overrideCaller) Nothing
                  assign (#state % #resetCaller) False
```

File: src/EVM.hs [hunk 4]
```diff
@@ -852,13 +852,18 @@ exec1 = do
...
+
+                newAddr <- createAddress from' this.nonce
```
Comment (elopez) [inline src/EVM.hs:862]: should the nonce be from's ? it might not be that important though if it's only off during pranking.

File: test/contracts/fail/10_BadVault.sol [hunk 5]
```diff
@@ -166,39 +175,4 @@ contract BadVaultTest is Test {
         // console2.log("attacker final balance", address(attacker).balance);
         assert(attacker.balance <= STARTING_BALANCE);
     }
-
```
Comment (elopez) [inline test/contracts/fail/10_BadVault.sol:169]: why is the test being removed?

File: test/test.hs [hunk 6]
```diff
@@ -1231,7 +1231,36 @@ tests = testGroup "hevm"
...
```
Comment (elopez) [inline test/test.hs:1256]: can we add a check to ensure the reset works correctly on creates?

```suggestion
                  assert(target.owner() == usr);
                  target = new Owned();
                  assert(target.owner() == address(this));
```

File: test/test.hs [hunk 7]
```diff
@@ -1483,7 +1512,7 @@ tests = testGroup "hevm"
     , test "Loop-Tests" $ do
         let testFile = "test/contracts/pass/loops.sol"
         runSolidityTestCustom testFile "prove_loop" Nothing (Just 10) False Nothing Foundry  >>= assertEqualM "test result" True
-        runSolidityTestCustom testFile "prove_loop" Nothing (Just 100) False Nothing Foundry >>= assertEqualM "test result" False
+        runSolidityTestCustom testFile "prove_loop" Nothing (Just 101) False Nothing Foundry >>= assertEqualM "test result" False
```
Comment (elopez) [inline test/test.hs:1515]: is this related to this PR?

File: test/contracts/pass/cheatCodes.sol [hunk 8]
```diff
@@ -65,8 +65,10 @@ contract CheatCodes is Test {
 
     function prove_warp_symbolic(uint128 jump) public {
         uint pre = block.timestamp;
-        hevm.warp(block.timestamp + jump);
-        assertEq(block.timestamp, pre + jump);
+        unchecked {
+          hevm.warp(block.timestamp + jump);
+          assertEq(block.timestamp, pre + jump);
+        }
```
Comment (elopez) [inline test/contracts/pass/cheatCodes.sol:71]: are these changes related to this PR?

File: test/contracts/fail/10_BadVault.sol (hunk 5 above)
Comment (msooseth) [reply test/contracts/fail/10_BadVault.sol:169]: No idea. I didn't do it :D I can add it back.

Comment (msooseth) [reply test/contracts/fail/10_BadVault.sol:169]: OK, added back the test!

File: test/contracts/pass/cheatCodes.sol (hunk 8 above)
Comment (msooseth) [reply test/contracts/pass/cheatCodes.sol]: Nope, it's because of `--via-ir`, which is not needed.

File: test/test.hs (hunk 7 above)
Comment (msooseth) [reply test/test.hs]: Nope, due to `--via-ir`, removing! Thanks for this!

File: src/EVM.hs (hunk 4 above)
Comment (msooseth) [reply src/EVM.hs:864]: I am actually not sure. This: https://book.getfoundry.sh/cheatcodes/prank says it: `Sets msg.sender to the specified address for the next call.` does not talk about nonce. No idea what that means.

**Review by msooseth — COMMENTED** [review]
@elopez thanks for the feedback! I made it a lot more lean :) I am unsure about the nonce thing. Do you know how we could check?

File: src/EVM.hs (hunk 4 above)
Comment (msooseth) [reply src/EVM.hs:864]: OK, we decided to leave this be for the moment.
